Samples/outlook-check-item-categories/src/commands/commands.js

```diff
@@ -0,0 +1,280 @@
+   let categories = [];
+   detectedWords.forEach(word => {
+     categories.push(`Office Add-ins Sample: ${word}`);
+   });
+ 
+   return categories;
+ }
+ 
+ /**
+  * Check that the appropriate categories, based on detected keywords in the subject or body, are applied to the 
+  * message or appointment before it is sent.
+  * @param {Office.AddinCommands.Event} event The OnMessageSend or OnAppointmentSend event object.
+  * @param {string[]} detectedWords The keywords found in the message or appointment's subject or body.
+  */
+  function checkAppliedCategories(event, detectedWords) {
+   let options = { asyncContext: { callingEvent: event, keywordArray: detectedWords } };
+   Office.context.mailbox.item.categories.getAsync(options, asyncResult => {
+     let sendEvent = asyncResult.asyncContext.callingEvent;
+     
+     if (asyncResult.status === Office.AsyncResultStatus.Succeeded) {
+       let requiredCategories = getCategoryName(asyncResult.asyncContext.keywordArray);
+       let detectedCategories = asyncResult.value;
+       if (detectedCategories) {
+         let detectedCategoryNames = getCategoryProperty(detectedCategories, "displayName");
+         let missingCategories = getMissingCategories(requiredCategories, detectedCategoryNames);
+         if (missingCategories.length > 0) {
+           let message = `Don't forget to also add the following categories: ${missingCategories.join(", ")}`;
+           console.log(message);
+           sendEvent.completed({ allowEvent: false, errorMessage: message });
+           return;
+         }
+ 
+         sendEvent.completed({ allowEvent: true });
+       } else {
+         let message = `You must assign the following categories before your ${Office.context.mailbox.item.itemType} can be sent: ${requiredCategories.join(", ")}`;
+         console.log(message);
+         sendEvent.completed({ allowEvent: false, errorMessage: message });
+         return;
+       }
+     } else {
+       console.log(asyncResult.error.message);
+       sendEvent.completed({ allowEvent: false, errorMessage: "Failed to check categories applied to the item." });
+       return;
+     }
+   });
+ }
+ 
+ /**
+  * Get the names of the required categories still missing from the message or appointment.
+  * @param {string[]} requiredCategories The names of the categories required on the message or appointment before it can be sent.
+  * @param {string[]} appliedCategories The names of the categories that are currently applied to the message or appointment.
+  * @returns {string[]} The names of the categories that need to be applied to the message or appointment.
+  */
+ function getMissingCategories(requiredCategories, appliedCategories) {
+   let missingCategories = requiredCategories.filter(category => !appliedCategories.includes(category));
+   return missingCategories;
+ }
+ 
+ Office.actions.associate("onItemComposeHandler", onItemComposeHandler);
+ Office.actions.associate("onItemSendHandler", onItemSendHandler);
```

Comment:
If you opt to have separate handler functions listed in the manifest, then the internal change could look like:
```suggestion
 Office.actions.associate("onMessageComposeHandler", onItemComposeHandler);
 Office.actions.associate("onAppointmentComposeHandler", onItemComposeHandler);
 Office.actions.associate("onMessageSendHandler", onItemSendHandler);
 Office.actions.associate("onAppointmentSendHandler", onItemSendHandler);
```